### ISSUE TITLE
Ignoring bogus URLs

### ISSUE THREAD

**Issue by wbrowar** [putyourlightson/craft-blitz#55]
Hey Ben,

I've got another issue I'm hoping we can solve—a coding challenge, if you will 😉

Our site launched and while I wanted to make sure all of the pages of the site are cached, using `.*` for the URL pattern means that a lot of garbage URLs are getting cached and kept track of in the database (some URLs may be from bots, some are just legacy URLs, and some look like customers typing in the URLs incorrectly). Surprisingly it's not a lot of query string URLs, but lots of incorrect URLs.

I don't mind the pages getting cached as static files because it's more efficient to hit a static page than to waste PHP processing on the page load, but the issue is they get added to the database and that means they are filling up the database with junk; and when I do cache warming it wastes time on pages that I don't care about.

A lot of these hit my 404 template since many of the URLs don't resolve to pages, so what I'm wondering is if there's a way that I could drop a Twig tag onto my 404 template that could somehow tell Blitz to cache the page, but don't add it to the database and don't warm it later. Technically you could just add a feature that does something like this for the 404, but having a tag gives me more flexibility.

---

Another way to look at this is that the reason I'm using `.*` as the URL pattern is because I don't know all of the URIs that will be created for this site. Some section types follow a URI pattern that could be added to Blitz's settings, but some URLs are based purely on the domain + the slug (for example: https://cbna.com/insurance).

Since the slug is content managed, if the client chooses to change it (or add a new page), manually putting that URL into Blitz's settings would require the client to contact us to make that change. I could ask the client to manually add in URL patterns, but that would mean giving them access to Settings and that's not something I want to do either.

A solution that would work for this particular site could be that Blitz gives me the option to automatically look at a section and add all of its entries' URIs into the list of URL patterns. Then I could remove `.*` and just add in the other patterns I know I'll need.

---

Both of these things could be their own features, but if one solution is better for the way Blitz is architected then either one will probably solve my issue.

Thanks,
Will

**Comment by putyourlightson** [comment]
Hey Will, please note that Blitz does not cache any pages that do not return a 200 status code, so any 404 pages will _not_ be cached. Can you check that this is the case on your site and let me know if and how that affects your request?

**Comment by wbrowar** [comment]
Hi Ben, I won't be able to access the project in question until next week, but I'll confirm what's going on and get back to you. Without looking at the code, one theory is that because I'm also using Retour to handle redirects, I wonder if the URLs are getting redirected to pages and then caching because they return 200 instead of 404.

The more I think about this, I'd prefer the second solution above, though. I like the URI pattern as an easy way to figure out what should be included and excluded from the cache, however, that first segment of the URI is hard to manage when that's generated by the user (unless there's a way to do this with URI patterns and I haven't figured it out). Just like an SEO plugin could generate a sitemap from the URIs in a section, it would be great if Blitz could automatically update the URI patterns to include all of the URIs in a section.

**Comment by putyourlightson** [comment]
Let me know once you've had a chance to take a closer look. Like I said, only a page that is loaded with a 200 status code will be cached, which excludes 301/302 redirects too.

**Comment by wbrowar** [comment]
Hey Ben, I got a chance to take a look at the URLs and how they're resolving and it looks like the plugin is working as it should and it comes down to my setup. My issue was a route that let in more than it should have and then query strings.

I've locked down the route and I have one page that needs the query string cached as separate pages (a search results page), so I think I'll turn off caching of query strings and maybe handle the search pages through URI segments instead.

I still think it would be great to have a way to automatically set URI patterns from sections that have public URIs, but it looks like your setup here handles things nicely.

It might be good to have a way to include query strings into consideration for caching so I technically could say "If you see `search` in the query string, cache that, but otherwise follow the rule I set in settings".

Thanks for taking the time to take a look for me and to explain how it handles status codes.

Will
